### ISSUE TITLE
Integration with Typst project

### ISSUE THREAD

**Issue by OpenFoam-User** [penrose/penrose#1622]
Typst is a new markup-based typesetting system that is powerful and easy to learn.

* Project link: https://github.com/typst/typst

Typst doesn't have built-in support for diagrams yet. It would be a huge improvement if Penrose is integrated with Typst  to generate complex diagrams (e.g. for academics).

See for example this issue: https://github.com/typst/typst/issues/729

Thank you

**Comment by wodeni** [comment]
That'd be awesome! Some our discord members chatted about making a plugin for Obsidian and I started experimenting with it here: https://github.com/wodeni/obsidian-penrose-plugin. Would be great to have Typst integration too. I guess to start the discussion: what are the technical constraints for the integration? Not super familiar with their ecosystem and would love to learn more.

**Comment by Enter-tainer** [comment]
Typst has a wasm based plugin system. Where it can run wasm32-baremetal code and get the result. I compiled quickjs javascript engine to wasm and made the [jogs](https://github.com/typst/packages/tree/main/packages/preview/jogs/0.2.1) package. So basically, we can say **typst can run javascript**.

We already have [wavedrom](https://github.com/typst/packages/tree/main/packages/preview/wavy/0.1.1), [mathjax](https://github.com/typst/packages/tree/main/packages/preview/m-jaxon/0.1.0), [pintora](https://github.com/typst/packages/pull/338) running in typst(although they are somehow slow-ish). 

Per https://github.com/penrose/penrose/issues/1688 penpose is possible to run out of browser and produce svg string without much modification. I will try if it can run in typst.

**Comment by Enter-tainer** [comment]
penpose is not a pure js project: optimizer contains wasm. So i guess it is not possible to run it inside typst now

**Comment by wodeni** [comment]
> penpose is not a pure js project: optimizer contains wasm. So i guess it is not possible to run it inside typst now

as of https://github.com/penrose/penrose/pull/1654 the optimizer is not longer written is rust. the only non-js part is actually produced at runtime by the autodiff engine (https://github.com/rose-lang/rose), which produces wasm modules at runtime. does this prevent typst from running penrose?

**Comment by samestep** [comment]
@wodeni Just to clarify, Rose itself also includes Wasm (compiled from Rust), not just the Wasm it dynamically generates.

**Comment by wodeni** [comment]
> @wodeni Just to clarify, Rose itself also includes Wasm (compiled from Rust), not just the Wasm it dynamically generates.

Right, so if typst requires all deps to be pure js, then `rose` would not be one

**Comment by Enter-tainer** [comment]
> > penpose is not a pure js project: optimizer contains wasm. So i guess it is not possible to run it inside typst now
> 
> as of https://github.com/penrose/penrose/pull/1654 the optimizer is not longer written is rust. the only non-js part is actually produced at runtime by the autodiff engine (https://github.com/rose-lang/rose), which produces wasm modules at runtime. does this prevent typst from running penrose?

typst can run js through the embedded quickjs engine. and the code wasm bindgen generated is target to browser. so it's not easy to run penpose in qjs and typst.
